File: backend/src/consultations/services/consultations.service.ts
import { Injectable, NotFoundException, BadRequestException, ForbiddenException } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';
import { BookConsultationDto, UpdateConsultationDto, RescheduleConsultationDto } from '../dto';
import { ConsultationStatus } from '@prisma/client';

@Injectable()
export class ConsultationsService {
  constructor(private prisma: PrismaService) {}

  async book(userId: string, bookConsultationDto: BookConsultationDto) {
    const { doctorId, scheduledAt, duration = 30, type } = bookConsultationDto;

    // Verify doctor exists and is active
    const doctor = await this.prisma.doctor.findUnique({
      where: { id: doctorId },
      include: { availabilitySlots: { where: { isActive: true } } },
    });

    if (!doctor) {
      throw new NotFoundException('Doctor not found');
    }

    if (!doctor.isActive) {
      throw new BadRequestException('Doctor is not accepting consultations');
    }

    // Validate scheduled time is in the future
    const scheduledDate = new Date(scheduledAt);
    if (scheduledDate <= new Date()) {
      throw new BadRequestException('Scheduled time must be in the future');
    }

    // Check for conflicts (doctor already has a consultation at this time)
    const endTime = new Date(scheduledDate.getTime() + duration * 60000);
    const conflictingConsultation = await this.prisma.consultation.findFirst({
      where: {
        doctorId,
        scheduledAt: {
          gte: new Date(scheduledDate.getTime() - 2 * 60 * 60000), // 2 hours before
          lte: new Date(scheduledDate.getTime() + 2 * 60 * 60000), // 2 hours after
        },
        status: {
          in: [ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED, ConsultationStatus.IN_PROGRESS],
        },
      },
    });

    if (conflictingConsultation) {
      throw new BadRequestException('Doctor is not available at this time');
    }

    // Check if the time falls within doctor's availability
    const dayOfWeek = scheduledDate.getDay();
    const timeString = `${scheduledDate.getHours().toString().padStart(2, '0')}:${scheduledDate.getMinutes().toString().padStart(2, '0')}`;
    
    const isWithinAvailability = doctor.availabilitySlots.some((slot) => {
      return slot.dayOfWeek === dayOfWeek && slot.startTime <= timeString && slot.endTime >= timeString;
    });

    if (!isWithinAvailability) {
      throw new BadRequestException('Selected time is outside doctor availability hours');
    }

    // Create consultation
    return this.prisma.consultation.create({
      data: {
        userId,
        doctorId,
        scheduledAt: scheduledDate,
        duration,
        type,
        fee: doctor.consultationFee,
        status: ConsultationStatus.SCHEDULED,
      },
      include: {
        doctor: {
          select: {
            id: true,
            name: true,
            specialization: true,
            imageUrl: true,
          },
        },
        user: {
          select: {
            id: true,
            username: true,
            email: true,
          },
        },
      },
    });
  }

  async findAllForUser(userId: string) {
    return this.prisma.consultation.findMany({
      where: { userId },
      include: {
        doctor: {
          select: {
            id: true,
            name: true,
            specialization: true,
            imageUrl: true,
          },
        },
      },
      orderBy: {
        scheduledAt: 'desc',
      },
    });
  }

  async findAllForAdmin(systemId: string) {
    return this.prisma.consultation.findMany({
      where: {
        doctor: {
          systemId,
        },
      },
      include: {
        doctor: {
          select: {
            id: true,
            name: true,
            specialization: true,
          },
        },
        user: {
          select: {
            id: true,
            username: true,
            email: true,
          },
        },
      },
      orderBy: {
        scheduledAt: 'desc',
      },
    });
  }

  async findOne(id: string, userId?: string) {
    const consultation = await this.prisma.consultation.findUnique({
      where: { id },
      include: {
        doctor: {
          select: {
            id: true,
            name: true,
            specialization: true,
            bio: true,
            qualifications: true,
            experience: true,
            imageUrl: true,
          },
        },
        user: {
          select: {
            id: true,
            username: true,
            email: true,
            profileType: true,
            journeyType: true,
          },
        },
      },
    });

    if (!consultation) {
      throw new NotFoundException('Consultation not found');
    }

    // If userId provided, verify ownership
    if (userId && consultation.userId !== userId) {
      throw new ForbiddenException('You do not have access to this consultation');
    }

    return consultation;
  }

  async reschedule(id: string, userId: string, rescheduleDto: RescheduleConsultationDto) {
    const consultation = await this.findOne(id, userId);

    // Only allow rescheduling scheduled or confirmed consultations
    const allowedStatuses = [ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED];
    if (!allowedStatuses.includes(consultation.status as any)) {
      throw new BadRequestException('Cannot reschedule consultation in current status');
    }

    // Validate new time is in the future
    const newScheduledDate = new Date(rescheduleDto.scheduledAt);
    if (newScheduledDate <= new Date()) {
      throw new BadRequestException('New scheduled time must be in the future');
    }

    // Check for conflicts at new time
    const endTime = new Date(newScheduledDate.getTime() + consultation.duration * 60000);
    const conflictingConsultation = await this.prisma.consultation.findFirst({
      where: {
        doctorId: consultation.doctorId,
        id: { not: id },
        scheduledAt: {
          gte: new Date(newScheduledDate.getTime() - 2 * 60 * 60000),
          lte: new Date(newScheduledDate.getTime() + 2 * 60 * 60000),
        },
        status: {
          in: [ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED, ConsultationStatus.IN_PROGRESS],
        },
      },
    });

    if (conflictingConsultation) {
      throw new BadRequestException('Doctor is not available at the new time');
    }

    return this.prisma.consultation.update({
      where: { id },
      data: {
        scheduledAt: newScheduledDate,
        status: ConsultationStatus.SCHEDULED, // Reset to scheduled
      },
      include: {
        doctor: {
          select: {
            id: true,
            name: true,
            specialization: true,
            imageUrl: true,
          },
        },
      },
    });
  }

  async cancel(id: string, userId: string) {
    const consultation = await this.findOne(id, userId);

    if (consultation.status === ConsultationStatus.CANCELLED) {
      throw new BadRequestException('Consultation is already cancelled');
    }

    if (consultation.status === ConsultationStatus.COMPLETED) {
      throw new BadRequestException('Cannot cancel completed consultation');
    }

    return this.prisma.consultation.update({
      where: { id },
      data: {
        status: ConsultationStatus.CANCELLED,
      },
      include: {
        doctor: {
          select: {
            id: true,
            name: true,
            specialization: true,
          },
        },
      },
    });
  }

  async updateStatus(id: string, updateDto: UpdateConsultationDto) {
    const consultation = await this.findOne(id);

    return this.prisma.consultation.update({
      where: { id },
      data: updateDto,
      include: {
        doctor: {
          select: {
            id: true,
            name: true,
            specialization: true,
          },
        },
        user: {
          select: {
            id: true,
            username: true,
            email: true,
          },
        },
      },
    });
  }

  async getAvailableSlots(doctorId: string, date: string) {
    // Get doctor's availability
    const doctor = await this.prisma.doctor.findUnique({
      where: { id: doctorId },
      include: {
        availabilitySlots: {
          where: { isActive: true },
        },
      },
    });

    if (!doctor || !doctor.isActive) {
      throw new NotFoundException('Doctor not found or inactive');
    }

    const requestedDate = new Date(date);
    const dayOfWeek = requestedDate.getDay();

    // Get availability for that day
    const dayAvailability = doctor.availabilitySlots.filter((slot) => slot.dayOfWeek === dayOfWeek);

    if (dayAvailability.length === 0) {
      return [];
    }

    // Get existing consultations for that day
    const startOfDay = new Date(requestedDate);
    startOfDay.setHours(0, 0, 0, 0);
    const endOfDay = new Date(requestedDate);
    endOfDay.setHours(23, 59, 59, 999);

    const existingConsultations = await this.prisma.consultation.findMany({
      where: {
        doctorId,
        scheduledAt: {
          gte: startOfDay,
          lte: endOfDay,
        },
        status: {
          in: [ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED, ConsultationStatus.IN_PROGRESS],
        },
      },
      select: {
        scheduledAt: true,
        duration: true,
      },
    });

    // Generate available time slots (30-minute intervals)
    const availableSlots: string[] = [];

    for (const availability of dayAvailability) {
      const [startHour, startMinute] = availability.startTime.split(':').map(Number);
      const [endHour, endMinute] = availability.endTime.split(':').map(Number);

      let currentTime = new Date(requestedDate);
      currentTime.setHours(startHour, startMinute, 0, 0);

      const endTime = new Date(requestedDate);
      endTime.setHours(endHour, endMinute, 0, 0);

      while (currentTime < endTime) {
        // Check if this slot conflicts with existing consultations
        const slotTime = new Date(currentTime);
        const hasConflict = existingConsultations.some((consultation) => {
          const consultationStart = new Date(consultation.scheduledAt);
          const consultationEnd = new Date(consultationStart.getTime() + consultation.duration * 60000);
          return slotTime >= consultationStart && slotTime < consultationEnd;
        });

        // Only add if no conflict and time is in the future
        if (!hasConflict && currentTime > new Date()) {
          availableSlots.push(currentTime.toISOString());
        }

        // Move to next 30-minute slot
        currentTime.setMinutes(currentTime.getMinutes() + 30);
      }
    }

    return availableSlots;
  }
}

File: backend/app/domains/consultations/services/consultations_service.py
from typing import List, Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_
from fastapi import HTTPException, status
from datetime import datetime, timedelta

from app.shared.models import Consultation, Doctor, ConsultationType, ConsultationStatus
from app.shared.schemas.consultation import (
    DoctorResponse, BookConsultationRequest, ConsultationResponse, 
    RescheduleConsultationRequest, AvailableSlot
)


class ConsultationsService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def get_user_consultations(self, user_id: str, system_id: str) -> List[ConsultationResponse]:
        result = await self.db.execute(
            select(Consultation)
            .where(Consultation.user_id == user_id)
            .where(Consultation.system_id == system_id)
            .order_by(Consultation.scheduled_at.desc())
        )
        consultations = result.scalars().all()
        return [self._consultation_to_response(consultation) for consultation in consultations]

    async def book_consultation(self, data: BookConsultationRequest, user_id: str, system_id: str) -> ConsultationResponse:
        # Verify doctor exists and belongs to system
        doctor_result = await self.db.execute(
            select(Doctor)
            .where(Doctor.id == data.doctorId)
            .where(Doctor.system_id == system_id)
        )
        doctor = doctor_result.scalar_one_or_none()
        
        if not doctor:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Doctor not found"
            )

        # Check if time slot is available
        existing_consultation = await self.db.execute(
            select(Consultation)
            .where(Consultation.doctor_id == data.doctorId)
            .where(Consultation.scheduled_at == data.scheduledAt)
            .where(Consultation.status.in_([ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED]))
        )
        
        if existing_consultation.scalar_one_or_none():
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Time slot is not available"
            )

        # Create consultation
        consultation = Consultation(
            user_id=user_id,
            doctor_id=data.doctorId,
            system_id=system_id,
            consultation_type=data.consultationType,
            scheduled_at=data.scheduledAt,
            status=ConsultationStatus.SCHEDULED,
            notes=data.notes,
            symptoms=data.symptoms
        )

        self.db.add(consultation)
        await self.db.commit()
        await self.db.refresh(consultation)

        return self._consultation_to_response(consultation)

    async def reschedule_consultation(self, consultation_id: str, data: RescheduleConsultationRequest, user_id: str, system_id: str) -> ConsultationResponse:
        # Get existing consultation
        result = await self.db.execute(
            select(Consultation)
            .where(Consultation.id == consultation_id)
            .where(Consultation.user_id == user_id)
            .where(Consultation.system_id == system_id)
        )
        consultation = result.scalar_one_or_none()
        
        if not consultation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Consultation not found"
            )

        if consultation.status not in [ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Cannot reschedule consultation in current status"
            )

        # Check if new time slot is available
        existing_consultation = await self.db.execute(
            select(Consultation)
            .where(Consultation.doctor_id == consultation.doctor_id)
            .where(Consultation.scheduled_at == data.newScheduledAt)
            .where(Consultation.id != consultation_id)
            .where(Consultation.status.in_([ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED]))
        )
        
        if existing_consultation.scalar_one_or_none():
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="New time slot is not available"
            )

        # Update consultation
        consultation.scheduled_at = data.newScheduledAt
        consultation.updated_at = datetime.now()

        await self.db.commit()
        await self.db.refresh(consultation)

        return self._consultation_to_response(consultation)

    async def cancel_consultation(self, consultation_id: str, user_id: str, system_id: str) -> None:
        result = await self.db.execute(
            select(Consultation)
            .where(Consultation.id == consultation_id)
            .where(Consultation.user_id == user_id)
            .where(Consultation.system_id == system_id)
        )
        consultation = result.scalar_one_or_none()
        
        if not consultation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Consultation not found"
            )

        if consultation.status in [ConsultationStatus.COMPLETED, ConsultationStatus.CANCELLED]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Cannot cancel consultation in current status"
            )

        consultation.status = ConsultationStatus.CANCELLED
        consultation.updated_at = datetime.now()

        await self.db.commit()

    async def get_available_slots(self, doctor_id: str, date: str) -> List[AvailableSlot]:
        # Parse date
        try:
            target_date = datetime.strptime(date, "%Y-%m-%d")
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid date format. Use YYYY-MM-DD"
            )

        # Verify doctor exists
        doctor_result = await self.db.execute(
            select(Doctor).where(Doctor.id == doctor_id)
        )
        doctor = doctor_result.scalar_one_or_none()
        
        if not doctor:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Doctor not found"
            )

        # Get existing consultations for the date
        start_of_day = target_date.replace(hour=0, minute=0, second=0, microsecond=0)
        end_of_day = start_of_day + timedelta(days=1)
        
        result = await self.db.execute(
            select(Consultation)
            .where(Consultation.doctor_id == doctor_id)
            .where(Consultation.scheduled_at >= start_of_day)
            .where(Consultation.scheduled_at < end_of_day)
            .where(Consultation.status.in_([ConsultationStatus.SCHEDULED, ConsultationStatus.CONFIRMED]))
        )
        existing_consultations = result.scalars().all()
        
        # Generate available slots (30-minute slots from 9 AM to 5 PM)
        available_slots = []
        current_time = start_of_day.replace(hour=9, minute=0)
        end_time = start_of_day.replace(hour=17, minute=0)
        
        booked_times = {c.scheduled_at for c in existing_consultations}
        
        while current_time < end_time:
            if current_time not in booked_times:
                available_slots.append(AvailableSlot(
                    startTime=current_time,
                    endTime=current_time + timedelta(minutes=30),
                    isAvailable=True
                ))
            current_time += timedelta(minutes=30)
        
        return available_slots

    def _consultation_to_response(self, consultation: Consultation) -> ConsultationResponse:
        return ConsultationResponse(
            id=consultation.id,
            doctorId=consultation.doctor_id,
            consultationType=consultation.consultation_type,
            scheduledAt=consultation.scheduled_at,
            status=consultation.status,
            notes=consultation.notes,
            symptoms=consultation.symptoms,
            createdAt=consultation.created_at,
            updatedAt=consultation.updated_at
        )
